lib/phoenix_html/form.ex

```diff
@@ -515,6 +515,59 @@ defmodule Phoenix.HTML.Form do
     generic_input(:range, form, field, opts)
   end
 
+  @doc """
+  Generates a date input.
+
+  Warning: this feature isn't available in all browsers.
+  Check `http://caniuse.com/#feat=input-datetime` for further informations.
+
+  See `text_input/3` for example and docs.
+  """
+  def date_input(form, field, opts \\ []) do
+    generic_input(:date, form, field, opts)
+  end
+
+  @doc """
+  Generates a datetime-local input.
+
+  Warning: this feature isn't available in all browsers.
+  Check `http://caniuse.com/#feat=input-datetime` for further informations.
+
+  See `text_input/3` for example and docs.
+  """
+  def datetime_local_input(form, field, opts \\ []) do
+    opts = Keyword.replace(opts, :value, datetime_local_input_value(Keyword.get(opts, :value)))
+    generic_input(:'datetime-local', form, field, opts)
+  end
+
+  defp datetime_local_input_value(%{__struct__: NaiveDateTime} = value) do
+    str = to_string(value)
+    String.slice(str, 0..9) <> "T" <> String.slice(str, 11..15)
+  end
+
+  defp datetime_local_input_value(other), do: other
+
+  @doc """
+  Generates a time input.
+
+  Warning: this feature isn't available in all browsers.
+  Check `http://caniuse.com/#feat=input-datetime` for further informations.
+
+  See `text_input/3` for example and docs.
+  """
+  def time_input(form, field, opts \\ []) do
+    opts = Keyword.replace(opts, :value, time_input_value(Keyword.get(opts, :value)))
```

Comment:
Please use `Keyword.put` so we remain compatible with earlier Elixir versions.

Reply:
I tried that first, but ran into this error (both for `time_input` and `datetime_local_input`:

```
$ mix test
Compiling 2 files (.ex)
.........................................................................................

  1) test datetime_local_input/3 with form (Phoenix.HTML.FormTest)
     test/phoenix_html/form_test.exs:358
     Assertion with == failed
     code:  assert safe_form(&(datetime_local_input(&1, :key))) == ~s"<input id=\"search_key\" name=\"search[key]\" type=\"datetime-local\" value=\"value\">"
     left:  "<input id=\"search_key\" name=\"search[key]\" type=\"datetime-local\">"
     right: "<input id=\"search_key\" name=\"search[key]\" type=\"datetime-local\" value=\"value\">"
     stacktrace:
       test/phoenix_html/form_test.exs:359: (test)
```

With, `Keyword.replace`, a `:value` entry is only added if it already existed before, while with `Keyword.put`, a `value: nil` is always added, overriding the value lookup from `form` in `generic_input` a few lines later:

```
Keyword.put_new(:value, input_value(form, field))
```

I couldn't find a way around this, other than adding lots of `if`s, which felt awkward. Would you have any backward-compatible suggestions?

Reply:
@josevalim This is one approach that works, but feels a bit verbose:

```
def time_input(form, field, opts \\ []) do
    opts = case Keyword.fetch(opts, :value) do
      {:ok, value} ->
        Keyword.put(opts, :value, time_input_value(value))
      :error ->
        opts
    end
    generic_input(:time, form, field, opts)
  end
```

Reply:
Yes, you will have to do the long version for now then.